File: java/nanocontainer/src/java/org/nanocontainer/integrationkit/LifecycleContainerBuilder.java
/*****************************************************************************
 * Copyright (C) NanoContainer Organization. All rights reserved.            *
 * ------------------------------------------------------------------------- *
 * The software in this package is published under the terms of the BSD      *
 * style license a copy of which has been included with this distribution in *
 * the LICENSE.txt file.                                                     *
 *                                                                           *
 *****************************************************************************/
package org.nanocontainer.integrationkit;

import org.picocontainer.MutablePicoContainer;
import org.picocontainer.PicoContainer;
import org.picocontainer.defaults.ObjectReference;

/**
 * @author <a href="mailto:joe@thoughtworks.net">Joe Walnes</a>
 * @author Aslak Helles&oslash;y
 * @version $Revision$
 */
public abstract class LifecycleContainerBuilder implements ContainerBuilder {

    public final void buildContainer(ObjectReference containerRef, ObjectReference parentContainerRef, Object assemblyScope) {
        PicoContainer parentContainer = parentContainerRef == null ? null : (PicoContainer) parentContainerRef.get();
        PicoContainer container = createContainer(parentContainer, assemblyScope);

        // register the child in the parent so that lifecycle can be propagated down the hierarchy
        if (parentContainer != null && parentContainer instanceof MutablePicoContainer) {
            MutablePicoContainer mutableContainer = (MutablePicoContainer) parentContainer;

            // this synchronization is necessary, because several servlet request may
            // occur at the same time for given session, and this produce race condition
            // especially in framed environments
            synchronized (mutableContainer) {
                mutableContainer.registerComponentInstance(containerRef, container);
            }
        }

        if (container instanceof MutablePicoContainer) {
            composeContainer((MutablePicoContainer) container, assemblyScope);
        }
        autoStart(container);

        // hold on to it
        containerRef.set(container);
    }

    protected void autoStart(PicoContainer container) {
        System.out.println("--> autostarted");
        container.start();
    }

    public void killContainer(ObjectReference containerRef) {
        try {
            PicoContainer pico = (PicoContainer) containerRef.get();
            pico.stop();
            pico.dispose();
            PicoContainer parent = pico.getParent();
            if (parent != null && parent instanceof MutablePicoContainer) {
                // see comment in buildContainer
                synchronized (parent) {
                    ((MutablePicoContainer) parent).unregisterComponentByInstance(pico);
                }
            }
        } finally {
            containerRef.set(null);
        }
    }

    protected abstract void composeContainer(MutablePicoContainer container, Object assemblyScope);

    protected abstract PicoContainer createContainer(PicoContainer parentContainer, Object assemblyScope);
}

File: dotnet/nanocontainer/trunk/container/src/NanoContainer/IntegrationKit/LifeCycleContainerBuilder.cs
using PicoContainer;
using PicoContainer.Defaults;

namespace NanoContainer.IntegrationKit
{
	public abstract class LifeCycleContainerBuilder : ContainerBuilder
	{
		private IMutablePicoContainer GetInstanceFromReference(IObjectReference objectReference)
		{
			return objectReference == null ? null : objectReference.Get() as IMutablePicoContainer;
		}

		public void BuildContainer(IObjectReference containerRef,
		                           IObjectReference parentContainerRef,
		                           object assemblyScope)
		{
			IMutablePicoContainer parent = GetInstanceFromReference(parentContainerRef);
			IMutablePicoContainer container = CreateContainer(parent, assemblyScope);

			// register the child in the parent so that lifecycle can be propagated down the hierarchy
			if (parent != null)
			{
				parent.UnregisterComponentByInstance(container);
				parent.RegisterComponentInstance(container, container);
			}

			ComposeContainer(container, assemblyScope);
			container.Start();

			containerRef.Set(container);
		}

		public void KillContainer(IObjectReference containerRef)
		{
			IMutablePicoContainer pico = GetInstanceFromReference(containerRef);

			try
			{
				if (pico != null)
				{
					pico.Stop();
					pico.Dispose();

					IMutablePicoContainer parent = pico.Parent as IMutablePicoContainer;
					if (parent != null)
					{
						parent.UnregisterComponentByInstance(pico);
					}
				}
			}
			finally
			{
				pico = null;
			}
		}

		protected abstract void ComposeContainer(IMutablePicoContainer container, object assemblyScope);

		protected abstract IMutablePicoContainer CreateContainer(IPicoContainer parentContainer, object assemblyScope);
	}
}